File: csharp/11CompleteImplementationCode/WordsFrequencyList.cs
using System;
using System.Collections.Generic;

namespace CompleteImplementationCode
{
	public class WordsFrequencyList {
		private const char COMMA_SEPARATOR = ',';

		public IList<WordEntity> CalculateWordsFrequency(string wordsList)
		{
			string[] splitWordsList = wordsList.Split(COMMA_SEPARATOR);

			var wordsFrequencyList = new List<WordEntity>();
			foreach (string word in splitWordsList)
			{
				string trimmedWord = word.Trim();
				if (string.IsNullOrEmpty(trimmedWord))
				{
					continue; // skip to the next word
				}
				int frequency = 1;
				WordEntity wordEntity = new WordEntity { Word = trimmedWord, Frequency = frequency };
				UpsertWordIntoWordsList(wordsFrequencyList, trimmedWord, wordEntity);
			}

			return wordsFrequencyList;
		}

		// upsert means update or insert
		private void UpsertWordIntoWordsList(IList<WordEntity> wordsFrequencyList, string word, WordEntity wordEntity)
		{
			WordEntity foundWord = FoundWordUsing(wordsFrequencyList, word);
			if (WordAlreadyExists(foundWord))
			{
				UpdateWordInWordsList(wordsFrequencyList, foundWord);
			}
			else
			{
				AddWordToWordsList(wordsFrequencyList, wordEntity);
			}
		}

		private void AddWordToWordsList(IList<WordEntity> wordsFrequencyList, WordEntity wordEntity) {
			wordsFrequencyList.Add(wordEntity);
		}

		private void UpdateWordInWordsList(IList<WordEntity> wordsFrequencyList, WordEntity foundWord) {
			wordsFrequencyList.Remove(foundWord);
			foundWord.IncrementFrequency();
			AddWordToWordsList(wordsFrequencyList, foundWord);
		}

		private bool WordAlreadyExists(WordEntity foundWord) {
			return foundWord != null;
		}

		private WordEntity FoundWordUsing(IList<WordEntity> wordEntityList, string word) {
			foreach (WordEntity wordEntity in wordEntityList)
			{
				if (wordEntity.isTheSameAs(word)) 
				{
					return wordEntity;
				}
			}
			return null;
		}
	}
}

File: java/11CompleteImplementationCode/src/main/java/com/neomatrix369/services/WordsFrequencyList.java
package com.neomatrix369.services;

import com.neomatrix369.domain.WordEntity;

import java.util.ArrayList;
import java.util.List;

public class WordsFrequencyList {
    private static final String COMMA_SEPARATOR = ",";

    public List<WordEntity> calculateWordsFrequency(String wordsList) {
        String[] splitWordsList = wordsList.split(COMMA_SEPARATOR);

        List<WordEntity> wordsFrequencyList = new ArrayList<>();
        for (String word: splitWordsList) {
            word = word.trim();
            if (word.isEmpty()) {
                continue; // skip to the next word
            }
            Integer frequency = 1;
            WordEntity wordEntity = new WordEntity(word, frequency);
            upsertWordIntoWordsList(wordsFrequencyList, word, wordEntity);
        }

        return wordsFrequencyList;
    }

    // upsert means update or insert
    private void upsertWordIntoWordsList(
            List<WordEntity> wordsFrequencyList,
            String word, WordEntity wordEntity) {
        WordEntity foundWord = foundWordUsing(wordsFrequencyList, word);
        if (wordAlreadyExists(foundWord)){
            updateWordInWordsList(wordsFrequencyList, foundWord);
        } else {
            addWordToWordsList(wordsFrequencyList, wordEntity);
        }
    }

    private void addWordToWordsList(List<WordEntity> wordsFrequencyList, WordEntity wordEntity) {
        wordsFrequencyList.add(wordEntity);
    }

    private void updateWordInWordsList(List<WordEntity> wordsFrequencyList, WordEntity foundWord) {
        wordsFrequencyList.remove(foundWord);
        foundWord.incrementFrequency();
        addWordToWordsList(wordsFrequencyList, foundWord);
    }

    private boolean wordAlreadyExists(WordEntity foundWord) {
        return foundWord != null;
    }

    private WordEntity foundWordUsing(
            List<WordEntity> wordEntityList, String word) {
        for (WordEntity wordEntity: wordEntityList) {
            if (wordEntity.isTheSameAs(word)) {
                return wordEntity;
            }
        }
        return null;
    }
}